client/src/includes/userbar.ts

```diff
@@ -11,6 +11,55 @@ This component implements a roving tab index for keyboard navigation
 Learn more about roving tabIndex: https://w3c.github.io/aria-practices/#kbd_roving_tabindex
 */
 
+export const sortAxeViolations = (violations: AxeResults['violations']) => {
+  interface inaccessibleNode {
+    selectedNode: HTMLElement;
+    selectorName: string;
+  }
+  const inaccessibleNodes: inaccessibleNode[] = [];
+
+  violations.forEach((violation) => {
+    violation.nodes.forEach((node) => {
+      const selectorName = node.target[0];
+      const selectedNode = document.querySelector<HTMLElement>(selectorName);
+      if (!selectedNode) return;
+      // Push both node (to compare DOM position) and selector (to sort Axe results)
+      inaccessibleNodes.push({
+        selectedNode,
+        selectorName,
+      });
+    });
+  });
+
+  const sortedNodes = inaccessibleNodes.sort((a, b) =>
+    // Method works with bitwise https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition
+    // eslint-disable-next-line no-bitwise
+    a.selectedNode.compareDocumentPosition(b.selectedNode) &
+    Node.DOCUMENT_POSITION_PRECEDING
+      ? 1
+      : -1,
+  );
+
+  // Utility object with index in order not to use find each time
+  const selectorIndex = {};
+  sortedNodes.forEach((sortedNode, index) => {
+    selectorIndex[sortedNode.selectorName] = index;
+  });
```

Comment:
Now that we have tests written, I think we could rewrite this to be clearer / more concise?

Another way we could write the algorithm for this code is:

- Sort all violations
  - Ordered based on position of their earliest DOM node
    - Which we can compute by sorting each violation’s nodes
      - Ordered based on comparing each node’s target position in the DOM

This way I think we’d be able to iterate much less over the nodes, getting rid of the `selectorIndex`, and `getMinIndex`. And potentially remove `inaccessibleNodes` as well – we can just iterate through each violation’s nodes the one time.